docs/develop/go/message-passing.mdx

```diff
@@ -27,135 +27,178 @@ tags:
+    ...
+    greeting := map[Language]string{English: "Hello", Chinese: "你好，世界"}
+    err := workflow.SetQueryHandler(ctx, GetLanguagesQuery, func(input GetLanguagesInput) ([]Language, error) {
+        // 👉 A Query handler returns a value: it can inspect but must not mutate the Workflow state.
+        if input.IncludeUnsupported {
+            return []Language{Chinese, English, French, Spanish, Portuguese}, nil
+        } else {
+            // Range over map is a nondeterministic operation.
+            // It is OK to have a non-deterministic operation in a query function.
+            //workflowcheck:ignore
+            return maps.Keys(greeting), nil
+        }
+    })
+    ...
 }
 ```
 
-### Handle a Signal {#handle-signal}
+- Use [`SetQueryHandler`](https://pkg.go.dev/go.temporal.io/sdk/workflow#SetQueryHandler) to set a Query Handler that listens for a Query by name.
+- Don't include any logic that causes [Command](/workflows#command) generation within a Query handler (such as executing Activities).
+Including such logic causes unexpected behavior.
+- The handler must be a function that returns two values, a serializable result and an error
 
-**How to handle a Signal using the Go SDK.**
+### Signal Channels {#handle-signal}
 
-Workflows listen for Signals by the Signal's name.
+A [Signal](/encyclopedia/workflow-message-passing#sending-signals) is a message sent asynchronously to a running Workflow Execution which can be used to change the state and control the flow of a Workflow Execution.
 
-Use the `GetSignalChannel()` API from the `go.temporal.io/sdk/workflow` package to get the Signal Channel.
+```go
+const ApproveSignal = "approve"
 
-A common use-case is to block a Workflow while waiting for a Signal, like in the following snippet:
+type ApproveInput struct {
+	Name string
+}
 
-```go
-func YourWorkflowDefinition(ctx workflow.Context, param YourWorkflowParam) error {
-  // ...
-  var signal MySignal
-  signalChan := workflow.GetSignalChannel(ctx, "your-signal-name")
-  signalChan.Receive(ctx, &signal)
-  if len(signal.Message) > 0 && signal.Message != "SOME_VALUE" {
-      return errors.New("signal")
-  }
-  // ...
+func GreetingWorkflow(ctx workflow.Context) error {
+    logger := workflow.GetLogger(ctx)
+	approverName := ""
+	...
+	// Block until the language is approved
+	var approveInput ApproveInput
+	workflow.GetSignalChannel(ctx, ApproveSignal).Receive(ctx, &approveInput)
+	approverName = approveInput.Name
+	logger.Info("Received approval", "Approver", approverName)
+	...
 }
 ```
 
-Alternatively, you might want the Workflow to proceed and still be capable of handling external Signals.
+- Use [`GetSignalChannel`](https://pkg.go.dev/go.temporal.io/sdk/workflow#GetSignalChannel) to get the Signal Channel that listen for Signals by the Signal's name.
+- Before completing the Workflow or using [Continue-As-New](/develop/go/continue-as-new), make sure to do an asynchronous drain on the Signal channel.
```

Comment:
This probably needs a link or a sample

Reply:
Agreed, @Quinn-With-Two-Ns what exactly is this instructing people to do? (Why "asynchronous"?)

Reply:
This was just taken from the existing message passing docs https://docs.temporal.io/develop/go/message-passing#handle-signal

You need to drain asynchronously because if you block while draining more signals can arrive. Here is a sample that does this https://github.com/temporalio/samples-go/blob/4b9e904654a0b369bf9b44f368117a1e86a918c9/batch-sliding-window/sliding_window_workflow.go#L195

Reply:
Thanks! I've added that snippet with a link to the sample. It needed explanation, and it's good to have some pointers here to more advanced example signal code.